### PR TITLE
Added the fix for column and table aliases contain single and multibyte characters.

### PR DESCRIPTION
**Pull request by riyajain39**
### Description

This pull request fixed the implementation for **column** and **table aliases** contain both single byte and multibyte characters whose length is more than 64 bytes for 14.latest version.

**Issue 1 :** Handle column and table aliases.
Column alias does not work when alias length is more than or equals to 64 bytes in case of multibyte aliases. When query select 1  as '您对“数据一览“中的车型，颜色，内饰，选装, '; executes through T-SQL endpoint, it causes client to hang and throws protocol error in TDS stream. The logic of [pre_transform_target_entry](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_3_X_DEV/contrib/babelfishpg_tsql/src/hooks.c#L1501C1-L1501C1) was not handling the aliases contain multibyte characters whose length is more than 64 bytes accurately and overwriting the alias name in case of truncation. To address this problem, I redefined the function and implemented the conditions which will handle aliases delimited with square bracket and double quotes separately with other aliases. In other aliases, I added the conditions which will handle both truncated and untruncated aliases separately.

**Issue 2 :** Aliases with multibyte chars cause MVU failure.
When establishing a view of alias with multibyte chars through T-SQL endpoint and retrieving it's definition from PG end point, extra unreadable chars are appearing at the end of the alias and even we can not use this view through tsql endpoint as it throws protocol error in TDS stream. But while we use similar alias on PG end point, we are able to retrieve it's correct definition. This discrepancy in behaviour is causing MVU failure. After implementing a fix, the correct alias definition is obtained from the PG endpoint, which rectifies the issue.

### Issues Resolved

BABEL-4231

Signed-off-by: Riya Jain <riyaajn@amazon.com>

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -1315,72 +1315,64 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
-			{
-				memcpy(alias + (NAMEDATALEN - 1) - 32,
-					   identifier_name + (NAMEDATALEN - 1) - 32,
-					   32);
-				alias[NAMEDATALEN] = '\0';
-			}
-
-			res->name = alias;
-		}
+        {
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Indentation seems odd

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 2]
```diff
@@ -1315,72 +1315,64 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+            bool        dq = *colname_start == '"';
+            bool        sqb = *colname_start == '[';
+            bool        sq = *colname_start == '\'';
+            bool        identifier_truncated;
+            const char *colname_end;
+
+            if (dq || sqb)
+            {
+                memcpy(alias, identifier_name, alias_len);
+            }
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: There's scope of optimisation here. We first read identifier_name from res->name and then copy content in alias and re-assign res->name with alias. In summation, this is kind of no-op. So just return from here in this case.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 3]
```diff
@@ -1315,72 +1315,64 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+            else
+            {
+                if(sq)
+                {
+                    colname_start++;
+                }
+                /*
+                * After truncation, maximum truncated length of alias can be 60
+                * If length is less than 60, it means alias is not truncated
+                */
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Fix indentation. Plus, we could elaborate more on this comment. We need to say that we need to handle identifier in special way only when truncated identifier length is b/w 61 to 63 to handle multibytes chars correctly.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 4]
```diff
@@ -1315,72 +1315,64 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+                else
+                {
+                    /*
+                    * To check whether last 32 bytes are equal to identifier_name or not.
+                    * If they are not equal, this means identifier_name is truncated
+                    */
+
+                    for(int x = alias_len - 32; x < alias_len; x++){
+                        colname_end = colname_start + x;
+                        if((*colname_end == identifier_name[x] || *colname_end == toupper(identifier_name[x])))
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Check if colname_end (also original char) is upper case then does uppercase of identifier_name matches? otherwise just compare colname_end with identifier_name.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 5]
```diff
@@ -1315,72 +1315,64 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+                        }
+                        else
+                        {
+                            identifier_truncated = true;
+                            memcpy(alias, colname_start, (alias_len - 32) );
+                            memcpy(alias + (alias_len + 1 - 1) - 32,
+                            identifier_name + (alias_len + 1 - 1) - 32, 
+                            32);
+                            alias[alias_len+1] = '\0';
+                            break;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Add comment to explain this. Also, `+ 1 - 1` is not needed just explain it in comments.

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 4 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 3 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 5 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 2 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 4 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/hooks.c]: This is not resolved.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/hooks.c]: Yet to be resolved

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 4 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved now.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Now resolved.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 6]
```diff
@@ -1316,71 +1316,113 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
 			colname_start = pstate->p_sourcetext + res->name_location;
 		}
 
+		/*
+         * Case 1 : Handle both singlebyte and multibyte aliases when delimited by
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Indentation:

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 7]
```diff
@@ -1316,71 +1316,113 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
-			bool		sq = *colname_start == '\'';
-			int			a = 0;
-			const char *colname_end;
-			bool		closing_quote_reached = false;
-
-			if (dq || sqb || sq)
-			{
-				colname_start++;
-			}
+        {
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: indentation

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 8]
```diff
@@ -1316,71 +1316,113 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
-					{
-						if ((*(++colname_end) != '"'))
-						{
-							closing_quote_reached = true;
-							break;	/* end of dbl-quoted identifier */
-						}
-					}
-					else if (sq && *colname_end == '\'')
+						/*
+                         * Check if colname_end is in upper case then does uppercase of identifier_name
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: indentation

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 9]
```diff
@@ -1316,71 +1316,113 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+            {
+                if(sq)
+                {
+                    colname_start++;
+                }
+                /*
+                 * After truncation, maximum truncated length of alias can be 60
+                 * If length is less than 60, it means alias is not truncated
+                 */
+                if(alias_len < 60){
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Open bracket should be on next line

Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 7 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 8 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

**Review by Deepesh125 — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 10]
```diff
@@ -1273,7 +1273,7 @@ find_attr_by_name_from_relation(Relation rd, const char *attname, bool sysColOK)
 
 static void
 pre_transform_target_entry(ResTarget *res, ParseState *pstate,
-						   ParseExprKind exprKind)
+						ParseExprKind exprKind)
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/hooks.c]: Unnecessary change

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 11]
```diff
@@ -1316,67 +1316,114 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+		 * more than or equals to 63 when not delimited by sqb and dq.
+		 * For instance, queries like : SELECT 1 AS 您对您对您对您对您对您对您对您对您对您对您对您对您对;
+		 */
 		if (alias_len > 0)
 		{
-			char	   *alias = palloc0(alias_len + 1);
-			bool		dq = *colname_start == '"';
-			bool		sqb = *colname_start == '[';
-			bool		sq = *colname_start == '\'';
-			int			a = 0;
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/hooks.c]: Unnecessary change

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 12]
```diff
@@ -1316,67 +1316,114 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
 			colname_start = pstate->p_sourcetext + res->name_location;
 		}
 
+		/*
+		 * Case 1 : Handle both singlebyte and multibyte aliases when delimited by 
+		 * square bracket(sqb) and double quoutes(dq).
+		 * For instance, queries like : SELECT 1 AS '您对“数据一览“中的车型，颜色，内饰，选装, ';
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/hooks.c]: The example is neither using square bracket nor using double quotes

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 13]
```diff
@@ -1316,67 +1316,114 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+		 * Case 2 : Preserve the case of aliases with ascii characters when there is no sqb and dq.
+		 * For instance, queries like: SELECT 1 AS ABCD;
+		 * Case 3 : Handle both singlebyte and multibyte aliases whose length is
+		 * more than or equals to 63 when not delimited by sqb and dq.
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/hooks.c:1328]: Why don't we have to worry about singlebyte and multibyte aliases whose length is more than or equals to 63 when not delimited by sqb and dq?

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 12 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 10 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 13 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c:1328]: Aliases when delimited with sqb and dq are already handled in grammer rule. These type of aliases are truncated properly and also the case is preserved. In such case, we are just returning. So now we need to handle aliases which are not delimited with sqb and dq whose length is less than 63 and more than 63. In case of single byte aliases, we need to preserve the case while in the case of multibyte aliases if actual alias length is more than 63, we should retrieve the alias name after truncation.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 11 above)
Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Did these changes to fix the indentation.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 14]
```diff
@@ -1318,67 +1318,114 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
...
+		 * For instance, queries like : SELECT 1 AS 您对您对您对您对您对您对您对您对您对您对您对您对您对;
+		 */
 		if (alias_len > 0)
 		{
 			char	   *alias = palloc0(alias_len + 1);
 			bool		dq = *colname_start == '"';
 			bool		sqb = *colname_start == '[';
 			bool		sq = *colname_start == '\'';
-			int			a = 0;
+			bool		identifier_truncated;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Dont leave this uninitialised. Initialise it with false

Comment (riyajain39) [reply contrib/babelfishpg_tsql/src/hooks.c]: Resolved.

**Review by Deepesh125 — APPROVED** [review]

**Review by kuntalghosh — APPROVED** [review]
